str2host.c

```diff
@@ -1800,3 +1800,282 @@ ldns_str2rdf_amtrelay(ldns_rdf **rd, const char *str)
+	if (strcmp(str, "port") == 0) { return 3; }
+	if (strcmp(str, "ipv4hint") == 0) { return 4; }
+	if (strcmp(str, "echconfig") == 0) { return 5; }
+	if (strcmp(str, "ipv6hint") == 0) { return 6; }
+	if (strncmp(str, "key", 3) == 0) {
+		int n = atoi(str + 3);
+		if (n > 0 && n <= USHRT_MAX) {
+			return n;
+		}
+	}
+	int n = atoi(str);
+	if (n > 0 && n <= USHRT_MAX) {
+		return n;
+	}
+	return -1;
+} 
+
+ldns_status
+ldns_str2rdf_svcbparams(ldns_rdf **rd, const char *str)
+{
+	uint8_t *data = NULL;
+	uint8_t *datap = NULL;
+	ldns_status status = LDNS_STATUS_OK;
+	int params_len;
+	const char *delimiters = "\n\t ";
+	char *token = NULL;
+	ldns_buffer *str_buf = NULL;
+	char *key_str = NULL;
+	char *val_str = NULL;
+	char *item_str = NULL;
+	uint8_t *item_buffer = NULL;
+	ldns_rdf *item_rdf = NULL;
+
+	if(rd == NULL) {
+		return LDNS_STATUS_NULL;
+	}
+
+	str_buf = LDNS_MALLOC(ldns_buffer);
+	if(!str_buf) {
+		return LDNS_STATUS_MEM_ERR;
+	}
+	ldns_buffer_new_frm_data(str_buf, (char *)str, strlen(str));
+	if(ldns_buffer_status(str_buf) != LDNS_STATUS_OK) {
+		status = LDNS_STATUS_MEM_ERR;
+		goto error;
+	}
+
+	token = LDNS_XMALLOC(char, LDNS_MAX_RDFLEN);
+	if (!token) {
+		status = LDNS_STATUS_MEM_ERR;
+		goto error;
+	}
+
+	data = LDNS_XMALLOC(uint8_t, LDNS_MAX_RDFLEN);
+	if (!data) {
+		status = LDNS_STATUS_MEM_ERR;
+		goto error;
+	}
+	datap = data;
+
+	while (ldns_bget_token(str_buf, token, delimiters, LDNS_MAX_RDFLEN) > 0) {
+		//printf("AWS token %s\n", token);
+		key_str = LDNS_XMALLOC(char, strlen(token) + 1);
+		if (!key_str) {
+			status = LDNS_STATUS_MEM_ERR;
+			goto error;
+		}
+		val_str = LDNS_XMALLOC(char, strlen(token) + 1);
+		if (!val_str) {
+			status = LDNS_STATUS_MEM_ERR;
+			goto error;
+		}
+		char *equals_in_str = strchr(token, '=');
+		if (equals_in_str == NULL) {
+			// Just a key with no value, set value str to empty string
+			strcpy(key_str, token);
+			val_str[0] = '\0';
+		} else {
+			uint equals_idx = equals_in_str - token;
+			uint token_len = strlen(token);
+			if (equals_idx > 0 && equals_idx < strlen(token) - 1) {
+				// '=' is in the middle of the string, split key and value on it
+				memcpy(key_str, token, equals_idx);
+				key_str[equals_idx] = '\0';
+				// Strip off qoutes around the value if they are used
+				if (token[equals_idx+1] == '"' && token[token_len-1] == '"') {
+					if (token_len - equals_idx < 4) {
+						status = LDNS_STATUS_INVALID_SVCB_PARAM_VALUE;
+						goto error;
+					}
+					memcpy(val_str, equals_in_str + 2, token_len - equals_idx - 2);
+					val_str[token_len - equals_idx - 3] = '\0';
+				} else {
+					strcpy(val_str, equals_in_str + 1);
+				}
+			} else {
+				// '=' is at the beginning or end of the token, this is not allowed
+				status = LDNS_STATUS_INVALID_SVCB_PARAM_VALUE;
+				goto error;
+			}
+		}
+
+		int key = ldns_svcbparams_text2key(key_str);
+		if (key == -1) {
+			status = LDNS_STATUS_INVALID_SVCB_PARAM_KEY;
+			goto error;
+		}
+		ldns_write_uint16(datap, key);
```

Comment:
This should be fixed (along with checking for duplicate keys), but will be a more involved change than the other fixes.  This will be done in a subsequent commit.